### PR TITLE
Add SPNEGO proxy authentication

### PR DESCRIPTION
**Pull request by kgraney**
Added support for Proxy-Authenticate and Proxy-Authorization headers,
which are used for proxy negotiation.  See RFC 2617.

### CODE REVIEW

File: spnego/spnego_server.go [hunk 1]
```diff
@@ -19,6 +19,9 @@ type ServerNegotiator interface {
 
 	// Negotiate handles the negotiation with the client.
 	Negotiate(*gssapi.CredId, http.Header, http.Header) (string, int, error)
+
+	// Negotiate handles the negotiation with the proxy client.
```
Comment (alextoombs) [inline spnego/spnego_server.go]: Minor: method comments should begin with the method name, so change this to `ProxyNegotiate`.

File: spnego/spnego_server.go [hunk 2]
```diff
@@ -54,20 +57,23 @@ func (k KerberizedServer) AcquireCred(serviceName string) (*gssapi.CredId, error
...
-// challenge that we send.
-func (k KerberizedServer) Negotiate(cred *gssapi.CredId, inHeader, outHeader http.Header) (string, int, error) {
-	negotiate, inputToken := CheckSPNEGONegotiate(k.Lib, inHeader, "Authorization")
+type ChallengeType struct {
+	ChallengeHeader string
+	ChallengeStatus int
+	AuthHeader      string
+}
+
+func (k KerberizedServer) DoNegotiation(t ChallengeType, cred *gssapi.CredId, inHeader,
```
Comment (alextoombs) [inline spnego/spnego_server.go]: Considering this is wrapped internally and not defined on the interface, I'd suggest not exporting the method.

**Comment by alextoombs** [comment]
So our usage of this is within our authentication service, where we implement the `ServerNegotiator` interface for handling the SPNEGO process.

As a consumer of this, I'm not a huge fan of adding an additional method to the interface; I like small interfaces, and this requires changes in our services in order to prevent compilation errors.

Is there any way we can maintain the original interface while inspecting the original incoming `http.Header`? Hiding the details of the proxy/non-proxy negotiation within the one exposed method?

**Comment by kgraney** [comment]
I don't think we can rely on the incoming HTTP header to indicate if it's a proxy connection or not.  The server is the first to respond with an authentication challenge, so it needs to initialize the conversation with Proxy-Authenticate.

We can keep the same interface if we use Go's polymorphic composition behavior that's not quite inheritance.  See kgraney/gssapi@72f9541 for a proposal.  This has the unfortunate side-effect of making `KerberizedProxyServer` initialization look somewhat ugly.
```
spnego.KerberizedProxyServer{spnego.KerberizedServer{lib}}
```
I can't think of a better way to handle this while letting the initialization of `KerberizedServer` remain the same.

**Comment by alextoombs** [comment]
Makes sense about the header.

I prefer the second proposal on your fork, personally. Even the initialization, while verbose, is intuitive to me; you're putting a proxy in front of a server, basically.

**Comment by levb** [comment]
why not just add `useProxyAuthentication bool` to `KerberizedServer`? And then use the flag in `Negotiate`? Seems like would be the simplest, and the most idiomatic

**Comment by kgraney** [comment]
@levb, do you mean like this?
```
type KerberizedServer struct {
    *gssapi.Lib
    useProxyAuthentication bool
}
```
Won't this change the interface for existing users who initialize this as `KerberizedServer{lib}`?

**Comment by levb** [comment]
Sorry, meant exported of course, Capitalized. 

```type KerberizedServer struct {
    *gssapi.Lib
    UseProxyAuthentication bool
}
```

Prior initializations of `KerberizedServer{lib}` will no longer compile, and would need to be changed to `KerberizedServer{Lib:lib}` which IMO is acceptable. @alextoombs agree?

**Comment by alextoombs** [comment]
That's fine with me. Agree on the exported nature.

**Comment by kgraney** [comment]
I'm happy with that.  The latest commit here has the change.

I think Travis is broken with an infrastructure bug?

**Comment by levb** [comment]
@kgraney Sorry to be a pest, but may I ask you to a) maybe get rid of the `type challengeType` and replace it with 3 local variables? and b) squash when you are done?

Otherwise, LGTM once the travis issues are worked out. @alextoombs do you know what's up with that? And MANY THANKS for contributing!

**Comment by kgraney** [comment]
Done, and thanks for doing all the work to get this library working with go!

**Comment by alextoombs** [comment]
For the Travis issue: I think we need this: https://github.com/apcera/gssapi/pull/47

**Comment by alextoombs** [comment]
@kgraney: I'm very sorry for the delay, but if you can rebase on top of master, this issue should be fixed.

The code looks great; thanks for contributing! As soon as the builds come back green, I'll merge.

**Comment by kgraney** [comment]
Cool, thanks!  It looks like the tests pass now.

**Comment by alextoombs** [comment]
LGTM

**Pull request merged**